src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SealInternalTypes.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Resx = Microsoft.NetCore.Analyzers.MicrosoftNetCoreAnalyzersResources;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class SealInternalTypes : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1851";
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            Resx.CreateLocalizableResourceString(nameof(Resx.SealInternalTypesTitle)),
+            Resx.CreateLocalizableResourceString(nameof(Resx.SealInternalTypesMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeHidden_BulkConfigurable,
+            Resx.CreateLocalizableResourceString(nameof(Resx.SealInternalTypesDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false,
+            isReportedAtCompilationEnd: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var candidateTypes = PooledConcurrentSet<INamedTypeSymbol>.GetInstance(SymbolEqualityComparer.Default);
+            var baseTypes = PooledConcurrentSet<INamedTypeSymbol>.GetInstance(SymbolEqualityComparer.Default);
+
+            context.RegisterSymbolAction(context =>
+            {
+                var type = (INamedTypeSymbol)context.Symbol;
+
+                if (!IsExternallyVisible(type) && type.TypeKind is TypeKind.Class && !type.IsAbstract && !type.IsStatic && !type.IsSealed)
```

Comment:
Keep cheap checks first.

```suggestion
                if (!type.IsAbstract && !type.IsStatic && !type.IsSealed && type.TypeKind is TypeKind.Class && !IsExternallyVisible(type))
```